xarray/testing.py

```diff
@@ -148,6 +155,68 @@ def compat_variable(a, b):
         raise TypeError("{} not supported by assertion comparison".format(type(a)))
 
 
+def _format_message(x, y, err_msg, verbose):
+    diff = x - y
+    abs_diff = max(abs(diff))
+    rel_diff = "not implemented"
+
+    n_diff = int(np.count_nonzero(diff))
+    n_total = diff.size
+
+    fraction = f"{n_diff} / {n_total}"
+    percentage = float(n_diff / n_total * 100)
+
+    parts = [
+        "Arrays are not equal",
+        err_msg,
+        f"Mismatched elements: {fraction} ({percentage:.0f}%)",
+        f"Max absolute difference: {abs_diff}",
+        f"Max relative difference: {rel_diff}",
+    ]
+    if verbose:
+        parts += [
+            f" x: {x!r}",
+            f" y: {y!r}",
+        ]
+
+    return "\n".join(parts)
+
+
+def assert_duckarray_allclose(
+    actual, desired, rtol=1e-07, atol=0, err_msg="", verbose=True
+):
+    """ Like `np.testing.assert_allclose`, but for duckarrays. """
+    __tracebackhide__ = True
+
+    allclose = duck_array_ops.allclose_or_equiv(actual, desired, rtol=rtol, atol=atol)
+    assert allclose, _format_message(actual, desired, err_msg=err_msg, verbose=verbose)
+
+
+def assert_duckarray_equal(x, y, err_msg="", verbose=True):
+    """ Like `np.testing.assert_array_equal`, but for duckarrays """
+    __tracebackhide__ = True
+
+    def array_like(x):
+        return hasattr(x, "ndim") and hasattr(x, "shape") and hasattr(x, "dtype")
+
+    def scalar(x):
+        return isinstance(x, (bool, int, float, complex, str)) or (
+            array_like(x) and x.ndim == 0
+        )
+
+    if not array_like(x) and not scalar(x):
+        x = np.asarray(x)
+
+    if not array_like(y) and not scalar(y):
+        y = np.asarray(y)
+
+    if (array_like(x) and scalar(y)) or (scalar(x) and array_like(y)):
+        equiv = (x == y).all()
```

Comment:
I'm surprised we want `assert_duckarray_equal(np.array([1, 1]), 1) == True`.

Reply:
well, that's what `numpy.testing.assert_array_equal` does. Even `numpy.testing.assert_equal` works that way (I'm not sure what the difference between both is, though)

Reply:
:+1: